src/background/modules/AutocorrectHandler.js

```diff
@@ -26,6 +26,76 @@ let antipatterns = [];
 // Adapted from: https://github.com/mozilla/webextension-polyfill/blob/master/src/browser-polyfill.js
 const IS_CHROME = Object.getPrototypeOf(browser) !== Object.prototype;
 
+/**
+ * Traverse Trie tree of objects to create RegEx.
+ *
+ * @param {Object.<string, Object|boolean>} obj
+ * @returns {string}
+ */
+function pattern(obj) {
+    const alt = [];
+    const cc = [];
+
+    // Escape special characters
+    const regExSpecialChars = /[.*+?^${}()|[\]\\]/gu;
+
+    for (const char in obj) {
+        if (char) {
+            const achar = char.replace(regExSpecialChars, "\\$&");
+
+            if (!("" in obj[char] && Object.keys(obj[char]).length === 1)) {
+                const recurse = pattern(obj[char]);
+                alt.push(recurse + achar);
+            } else {
+                cc.push(achar);
+            }
+        }
+    }
+
+    if (cc.length) {
+        alt.push(cc.length === 1 ? cc[0] : `[${cc.join("")}]`);
+    }
+
+    let result = alt.length === 1 ? alt[0] : `(?:${alt.join("|")})`;
+
+    if ("" in obj) {
+        if (cc.length || alt.length > 1) {
+            result += "?";
+        } else {
+            result = `(?:${result})?`;
+        }
+    }
+
+    return result;
+}
+
+/**
+ * Convert autocorrections into Trie tree of objects.
+ *
+ * @param {string[]} arr
+ * @returns {string}
+ */
+function createRegEx(arr) {
+    const tree = {};
+
+    for (const s of arr.sort((a, b) => b.length - a.length)) {
+        let temp = tree;
+
+        for (const char of Array.from(s).reverse()) {
+            if (!(char in temp)) {
+                temp[char] = {};
+            }
+            temp = temp[char];
+        }
+
+        // Leaf node
+        temp[""] = true;
+    }
+
+    Object.freeze(tree);
+    return pattern(tree);
```

Comment:
Also I am generally confused here: You take the three and then anyway make a RegEx as it was before? How/Why is this better/optimized than before?

Reply:
Before it would generate a long RegEx like this for example:
```js
/foobar|foobah|fooxar|foozap|fooza/
```
Using the Trie tree it can now generate a more optimized one:
```js
/foo(?:ba[rh]|xar|zap?)/
```
This should both reduce memory usage (it is shorter) and improve the performance of the autocorrection feature (it does not have to do costly backtracking to check for a match). Note that the resulting RegEx is stored in the content script on every frame of every page, so small reductions in size will make a big difference in memory usage.

For Unicodify specifically, the before is this:
`/(--|\(e\)|\.\.\.|::|:=|~=|\+-|-\+|<<|<=|->|>=|>>|<=>|1/4|1/2|3/4|1/7|1/9|1/10|1/3|2/3|1/5|2/5|3/5|4/5|1/6|5/6|1/8|3/8|5/8|7/8|<--|<==|==>|-->|<-->|<--\||<==\||\|-->|\|==>|::=|__|==|!=|===|<-<|<<-|<-|<->|->>|>->|<==>|=>|>>-|>-|<~>|-<|-<<|<-\||<=\||/\\|\\/|\|->|\|=>|<~~|<~|~~|~>|~~>|<<<|<>|>>>|<\||\|>|\\\\|\|\||\.\.|---|\+\+|\+\+\+|-\||_\|_|\|-|\|\|-|\|=|\|\|=)$/`
and the after is this:
`/(1/10|1/2|[12]/3|[13]/4|[1234]/5|[15]/6|1/7|[1357]/8|1/9|(?:(?:[<\|]?-|[<>\|])?-|(?:[\|<]?=|[<\|])?=|[->]?>|[<~]?~|[<\|])>|(?:(?:<-|<)?-|(?:<=|<)=|[<\|])\||\(e\)|\.?\.\.|(?:[<-]?-|<?<|>?>|\|?\||\+)-|(?:[<=]?=|:?:|\|?\||[~<>!])=|(?:<?-|[-<]?<)<|(?:<?~|<)~|(?:\+?\+|-)\+|(?:_\||_)_|::|[/\\]\\|\\/)$/`

It will of course make a much bigger difference for the Awesome Emoji Picker where there are thousands of supported autocorrections (at least one for every emoji) and many of them are much longer. They are way too long to post here in full, but here is the beginning of each:
![image](https://user-images.githubusercontent.com/15305150/209470014-e9709e28-d2d2-447f-81d8-e71ed25611f4.png)
The before (on top) is 24,025 characters and after (on bottom) is 18,501.

Reply:
Thanks a lot! That makes sense.